File: coding/src/com/algos/core/li10_tree/bt/Bt_L107_BinaryTreeLevelOrderTraversalII.java
package  com.algos.core.li10_tree.bt;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

import  com.algos.core.li30_model.TreeNode;

public class Bt_L107_BinaryTreeLevelOrderTraversalII {
    public static void main(String[] args) {
        Bt_L107_BinaryTreeLevelOrderTraversalII obj = new Bt_L107_BinaryTreeLevelOrderTraversalII();

        		TreeNode root = new TreeNode(1);
		root.left = new TreeNode(3);
		root.right = new TreeNode(4);
		root.left.left = new TreeNode(2);
        root.left.right = new TreeNode(6);
        root.left.right.left = new TreeNode(5);

        System.out.println(obj.levelOrderBottom(root));;
    }

    // Problem: https://leetcode.com/problems/binary-tree-level-order-traversal-ii/
    // Idea: Need 2 queues if you want to print level by level.
    // Instead of using the 2nd queue, rely on 1st queue size and process the queue element until its size. Keep insert elements into same queue.
    // We have to insert the level list at the begin of linked list as want them in reverse order.
    // Time: O(n)
    // space: O(n)
    public List<List<Integer>> levelOrderBottom(TreeNode root) {
        LinkedList<List<Integer>> result = new LinkedList<List<Integer>>();
        if(root == null)
            return result;
            
        Queue<TreeNode> q = new ArrayDeque<TreeNode>();
		q.add(root);
		while(!q.isEmpty()) {
            List<Integer> level = new ArrayList<Integer>();
            int size = q.size();
			for(int i=0; i<size; i++) {
				TreeNode curr = q.remove();
				level.add(curr.val);
                if(curr.left != null)
                    q.add(curr.left);    
                if(curr.right != null)
                    q.add(curr.right);
			}

            result.addFirst(level);    
		}

        return result;
    }
}

File: dsa/concepts/10_trees/bt/python/bt_l107__binary_tree_level_order_traversal_ii.py
from collections import deque

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# Problem: https://leetcode.com/problems/binary-tree-level-order-traversal-ii/
# Idea: Process queue level by level. Insert each level at the beginning of result for bottom-up order.
# Time: O(n)
# Space: O(n)

def level_order_bottom(root):
    result = deque()  # Use deque to appendleft efficiently
    if not root:
        return []

    q = deque([root])

    while q:
        level = []
        size = len(q)
        for _ in range(size):
            curr = q.popleft()
            level.append(curr.val)
            if curr.left:
                q.append(curr.left)
            if curr.right:
                q.append(curr.right)
        result.appendleft(level)  # Add at beginning for bottom-up order

    return list(result)  # Convert deque back to list


# -------- Test --------
root = TreeNode(1)
root.left = TreeNode(3)
root.right = TreeNode(4)
root.left.left = TreeNode(2)
root.left.right = TreeNode(6)
root.left.right.left = TreeNode(5)

print(level_order_bottom(root))  
# Expected: [[5], [2, 6], [3, 4], [1]]
